### PR TITLE
[Feature] New Codec for IPR, a wrapper for multi task losses

### PR DESCRIPTION
**Pull request by Tau-J**
## Motivation

- implement a new Codec for IPR method
- rename original IPR as SoftArgmaxHead to perform saparate coord regression
- implement a wrapper for multi-task losses, which supports arbitrary number of losses
- implement a new IPR head 

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit tests to ensure correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by Tau-J** [comment]
I would like to use `multi_loss_factory`, but it is already used by the bottom-up method

File: configs/body_2d_keypoint/topdown_regression/coco/td-reg_res50_ipr-8xb64-210e_coco-256x192.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    type='TopdownPoseEstimator',
+    data_preprocessor=dict(
+        type='PoseDataPreprocessor',
+        mean=[123.675, 116.28, 103.53],
+        std=[58.395, 57.12, 57.375],
+        bgr_to_rgb=True),
+    backbone=dict(
+        type='ResNet',
+        depth=50,
+        init_cfg=dict(type='Pretrained', checkpoint='torchvision://resnet50'),
```
Comment (ly015) [inline configs/body_2d_keypoint/topdown_regression/coco/td-reg_res50_ipr-8xb64-210e_coco-256x192.py:46]: The COCO-pretrained model can be loaded from https://download.openmmlab.com/mmpose/pretrain_models/td-hm_res50_8xb64-210e_coco-256x192.pth

**Comment by ly015** [comment]
As for the organization of the new configs, I suggest we consider SimCC, IPR, and DSNT as algorithms parallel to top-down heatmap (td-hm) and top-down regression (td-reg, which particularly refers to deeppose or directly regressing coordinates with a FC layer).

**Comment by Tau-J** [comment]
> As for the organization of the new configs, I suggest we consider SimCC, IPR, and DSNT as algorithms parallel to top-down heatmap (td-hm) and top-down regression (td-reg, which particularly refers to deeppose or directly regressing coordinates with a FC layer).

Should we also reorganize the heads of new methods?

File: mmpose/models/losses/multi_task_loss.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch.nn as nn
+
+from mmpose.registry import MODELS
+
+
+@MODELS.register_module()
+class MultiTaskLoss(nn.Module):
```
Comment (ly015) [inline mmpose/models/losses/multi_task_loss.py:8]: I suggest we name this class 'CombinedLossWrapper' or 'MultipleLossWrapper' to show that it's actually a wrapper instead of a real loss function. And this file can be renamed 'loss_wrappers.py`.

File: mmpose/models/losses/multi_task_loss.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    """Combine multi-task losses together with different factors.
+
+    Args:
+        loss_cfg_list (list): List of Loss Config
+        factors (list): List of Loss factors
+    """
+
+    def __init__(self, loss_cfg_list: list, factors: list):
```
Comment (ly015) [inline mmpose/models/losses/multi_task_loss.py:16]: About the argument names, I think 'losses' and 'loss_weights' would be simpler and clearer.

File: mmpose/models/losses/multi_task_loss.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        for loss_cfg in loss_cfg_list:
+            t_loss = MODELS.build(loss_cfg)
+            losses.append(t_loss)
+        self.losses = nn.ModuleList(losses)
+
+    def forward(self,
+                input_list,
+                target_list,
+                keypoint_weights=None,
+                epoch_facotrs=None):
```
Comment (ly015) [inline mmpose/models/losses/multi_task_loss.py]: 'epoch_facotrs` seems a typo.

More importantly, I think this epoch-based loss weight adjustment should be handled by the head instead of the loss since it's an algorithm-specific behavior. As a loss wrapper, it should have a consistent signature with other losses here.

File: configs/body_2d_keypoint/integral_regression/coco/resnet_ipr_coco.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
+<!-- [ALGORITHM] -->
+
+<details>
+<summary align="right"><a href="https://arxiv.org/abs/1711.08229">DSNT (ECCV'2018)</a></summary>
```
Comment (ly015) [inline configs/body_2d_keypoint/integral_regression/coco/resnet_ipr_coco.md:4]:
```suggestion
<summary align="right"><a href="https://arxiv.org/abs/1711.08229">IPR (ECCV'2018)</a></summary>
```

File: configs/body_2d_keypoint/integral_regression/coco/resnet_dsnt_coco.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+}
+```
+
+</details>
+
+Results on COCO val2017 with detector having human AP of 56.4 on COCO val2017 dataset
+
+| Arch                     | Input Size |  AP   | AP<sup>50</sup> | AP<sup>75</sup> |  AR   | AR<sup>50</sup> |    ckpt    |    log    |
+| :----------------------- | :--------: | :---: | :-------------: | :-------------: | :---: | :-------------: | :--------: | :-------: |
+| [ipr_resnet_50_dsnt](<>) |  256x256   | 0.667 |      0.868      |      0.736      | 0.759 |      0.926      | [ckpt](<>) | [log](<>) |
```
Comment (ly015) [inline configs/body_2d_keypoint/integral_regression/coco/resnet_dsnt_coco.md:56]: Please fill in the config paths. And I will update the checkpoint and log URLs after uploading the files.

Comment (Tau-J) [reply configs/body_2d_keypoint/integral_regression/coco/resnet_dsnt_coco.md]: done

File: configs/body_2d_keypoint/simcc/coco/resnet_coco.md [hunk 7]
```diff
@@ -4,13 +4,22 @@
...
-@misc{li20212d,
-      title={Is 2D Heatmap Representation Even Necessary for Human Pose Estimation?},
-      author={Yanjie Li and Sen Yang and Shoukui Zhang and Zhicheng Wang and Wankou Yang and Shu-Tao Xia and Erjin Zhou},
-      year={2021},
-      eprint={2107.03332},
-      archivePrefix={arXiv},
-      primaryClass={cs.CV}
+@misc{https://doi.org/10.48550/arxiv.2107.03332,
+  doi = {10.48550/ARXIV.2107.03332},
+
```
Comment (jin-s13) [inline configs/body_2d_keypoint/simcc/coco/resnet_coco.md:9]: Please delete this blank line

File: mmpose/models/heads/heatmap_heads/simcc_head.py [hunk 8]
```diff
@@ -275,18 +289,17 @@ def loss(
         )
 
         target_coords = np.concatenate(
-            [d.gt_instances.keypoints for d in batch_data_samples], axis=0)
+            [d.gt_instances.transformed_keypoints for d in batch_data_samples],
```
Comment (ly015) [inline mmpose/models/heads/heatmap_heads/simcc_head.py:292]: Why do we use `transformed_keypoints` instead of `keypoints` here? In the training pipeline `transformed_keypoints` will be packed as `keypoints` so they should be identical. Actually I think `transformed_keypoints` serves as an intermediate variable in pipelines and shouldn't be packed and used by models.

**Review by ly015 — APPROVED** [review]
